Source/DafnyRuntime/DafnyRuntime.go

```diff
@@ -692,183 +692,412 @@ func (seq Seq) VerbatimString(asLiteral bool) string {
-  arr := make([]interface{}, len(values))
+  // Use the default representation
+  arr := make([]interface{}, totalLength)
   copy(arr, values)
-  return &Array{
+  return &ArrayStruct{
     contents: arr,
-    dims:     []int{len(values)},
+    dims:     intDims,
   }
 }
 
-// Len returns the length of the array in the given dimension.
-func (array *Array) Len(dim int) Int {
-  return IntOf(array.LenInt(dim))
+// NewArrayWithValue returns a new Array full of the given initial value.
+func NewArrayWithValue(init interface{}, dims ...Int) Array {
+  return NewArrayFromExample(init, init, dims...)
 }
 
-// LenInt returns the length of the array in the given dimension, as an int.
-func (array *Array) LenInt(dim int) int {
-  return array.dims[dim]
+// NewArray returns a new Array full of the default value of the given type.
+func NewArray(dims ...Int) Array {
+  return NewArrayFromExample(nil, nil, dims...)
 }
 
-// Equals compares two arrays for equality.  Values are compared using
-// dafny.AreEqual.
-func (array *Array) Equals(array2 *Array) bool {
-  if array == array2 {
-    return true
+/***** ArrayStruct is default implementation of the Array interface. *****/
+
+type ArrayStruct struct {
+  contents []interface{} // stored as a flat one-dimensional slice
+  dims     []int
+}
+
+func (_this ArrayStruct) dimensionCount() int {
+  return len(_this.dims)
+}
+
+func (_this ArrayStruct) dimensionLength(dim int) int {
+  return _this.dims[dim]
+}
+
+func (_this ArrayStruct) ArrayGet1(index int) interface{} {
+  return _this.contents[index]
+}
+
+func (_this ArrayStruct) ArraySet1(value interface{}, index int) {
+  _this.contents[index] = value
+}
+
+func (_this ArrayStruct) ArrayGet1Byte(index int) byte {
+  panic("cannot specialize general ArrayGet with byte")
+}
+
+func (_this ArrayStruct) ArraySet1Byte(value byte, index int) {
+  panic("cannot specialize general ArraySet with byte")
+}
+
+func (_this ArrayStruct) ArrayGet1Char(index int) Char {
+  panic("cannot specialize general ArrayGet with char")
+}
+
+func (_this ArrayStruct) ArraySet1Char(value Char, index int) {
+  panic("cannot specialize general ArraySet with char")
+}
+
+func (_this ArrayStruct) anySlice(lo, hi Int) []interface{} {
+  if lo.IsNilInt() && hi.IsNilInt() {
+    return _this.contents
   }
-  if array == nil || array2 == nil {
-    return false // we already know they're not equal as pointers
+  if lo.IsNilInt() {
+    return _this.contents[:hi.Int()]
   }
-  if len(array.dims) != len(array2.dims) {
-    return false
+  if hi.IsNilInt() {
+    return _this.contents[lo.Int():]
   }
-  for i, d := range array.dims {
-    if d != array2.dims[i] {
-      return false
-    }
+  return _this.contents[lo.Int():hi.Int()]
+}
+
+// ArrayStruct implements the EqualsGeneric interface.
+func (_this ArrayStruct) EqualsGeneric(other interface{}) bool {
+  otherArray, ok := other.(*ArrayStruct)
+  if !ok {
+    return false
   }
+  return &_this.dims[0] == &otherArray.dims[0]
```

Comment:
Why not just `&_this == &otherArray`?

Reply:
Your proposed comparison (actually, it would be `&_this == otherArray` to have the right type) would compare the addresses of the two `ArrayStruct`s. When a Dafny array is copied, what happens in the target Go is that such an `ArrayStruct` is copied. Thus, the address of the two `ArrayStruct`s may be different and yet they might refer to the same array.

This important test case was missing from the test suite. I have added it, along with a comment next to L911.

Reply:
I'm still a bit confused. I understand you're saying that Go will copy the struct as it is passed around, and we want different structs that refer to the same underlying storage to compare equal. But why compare the address of the first element of the `dims` array specifically?

Reply:
I argued above that `&_this == otherArray` does not work. But then, what about

    _this.contents == otherArray.contents

or

    _this.dims == otherArray.dims

This doesn't work either, because the type of `.contents` and `.dims` is a _slice_ and the only thing Go allows you to compare a slice with is `nil`. One might imagine obtaining the Go arrays underlying these slices and comparing them. However, that would require unsafe Go code.

Taking the address of an array/slice element is always okay in Go. The specific element we're taking the address does not matter, as long as it's the same for the LHS and RHS of the comparison and as long as the element exists. Using the address of `.contents[0]` is _almost_ the obvious choice, except that the contents may be empty. However, the number of dimensions is always nonempty (so `.dims[0]` exists) and the implementation currently always uses a different `.dims` for each `Array`. (If that were to change in the future, to someone share these never-mutated `.dims` arrays, then the equality comparison we use will need to change.)

Reply:
I changed `ArrayStruct`, `ArrayForByte`, `ArrayForChar`, and `ArrayForCodePoint` to be private to the package in order to contain these design decisions in the Dafny package.